wgengine/netstack/link_endpoint.go

```diff
@@ -132,19 +156,117 @@ func (l *linkEndpoint) Drain() int {
 	return c
 }
 
-// NumQueued returns the number of packet queued for outbound.
+// NumQueued returns the number of packets queued for outbound.
 func (l *linkEndpoint) NumQueued() int {
 	return l.q.Num()
 }
 
-// InjectInbound injects an inbound packet. If the endpoint is not attached, the
-// packet is not delivered.
-func (l *linkEndpoint) InjectInbound(protocol tcpip.NetworkProtocolNumber, pkt *stack.PacketBuffer) {
+// rxChecksumOffload validates IPv4, TCP, and UDP header checksums in p,
+// returning an equivalent *stack.PacketBuffer if they are valid, otherwise nil.
+// The set of headers validated covers where gVisor would perform validation if
+// !stack.PacketBuffer.RXChecksumValidated, i.e. it satisfies
+// stack.CapabilityRXChecksumOffload.
+func rxChecksumOffload(p *packet.Parsed) *stack.PacketBuffer {
+	var (
+		pn        tcpip.NetworkProtocolNumber
+		csumStart int
+	)
+	buf := p.Buffer()
+
+	switch p.IPVersion {
+	case 4:
+		if len(buf) < 20 {
+			return nil
+		}
+		csumStart = int((buf[0] & 0x0F) * 4)
+		if csumStart < 20 || csumStart > 60 || len(buf) < csumStart {
+			return nil
+		}
+		if ^tun.Checksum(buf[:csumStart], 0) != 0 {
+			return nil
+		}
+		pn = header.IPv4ProtocolNumber
+	case 6:
+		if len(buf) < 40 {
+			return nil
+		}
+		csumStart = 40
+		pn = header.IPv6ProtocolNumber
+	}
+
+	if p.IPProto == ipproto.TCP || p.IPProto == ipproto.UDP {
```

Comment:
are other protos handled correctly?

Reply:
Other protos don't require offload, e.g. ICMP header checksum is still validated inside gVisor when `RXChecksumValidated == true`. This is in the function docstring, but I've added more there.

https://github.com/tailscale/tailscale/pull/12921/commits/e5622bde88db9258de971a0a0cbfbb35e2d670ce

Reply:
IPv6 extension headers require a little thought here. `packet.Parsed` does not support parsing them, so we will get `p.IPProto == 0` when say there is an extension header, but then TCP header following the extension header.

Reply:
I was mostly thinking SCTP which we have occasionally needed to fix up since it's semi-supported.

IPv6 extension headers are a bit of a pain, but I think their existence normally causes Linux to fall out of the offload fast paths? Or maybe that was just if the extension headers differ from packet to packet.

Reply:
SCTP isn't supported in gVisor, and there's no plans to add it based on https://github.com/google/gvisor/issues/5000#issuecomment-755045332, so we are ok there for now.

The case I'm concerned about is `IPv6Header | IPv6ExtHdr | TCP/UDPHeader` where this offload logic will not currently validate the inner TCP/UDP header checksum, but its marked as such. I need to look closer at gVisor and see what support it has for extension headers and how it will behave in that case.

Reply:
gVisor does support extension header parsing, and the presence of extension headers doesn't change any conditions around rx checksum offloading, i.e. we will not validate the L4 checksum, but gVisor will assume we did.

I can either add a TODO, or add a slower path where we use gVisor parsing and then validate. The former may be appropriate since we don't support extension headers in `packet.Parsed`, which can breaks other things.

Thoughts?

Reply:
Probably better to add the slow path rather than getting confusing problem reports about failures due to extra headers. 

I think extension headers are fairly normal at the start of a connection?

Reply:
> Probably better to add the slow path rather than getting confusing problem reports about failures due to extra headers.

Agreed. Implementation in https://github.com/tailscale/tailscale/pull/12921/commits/00a5b53c573cad9892aaac93162de1d5777a0e63. Once that fixup is reviewed this can be merged.

> I think extension headers are fairly normal at the start of a connection?

I think usage is fairly limited.